packages/eslint-plugin/src/rules/init-declarations.ts

```diff
@@ -28,7 +28,51 @@ export default createRule<Options, MessageIds>({
   },
   defaultOptions: ['always'],
   create(context, [mode]) {
-    const rules = baseRule.create(context);
+    // Make a custom context to adjust the the loc of reports where the base
+    // rule's behavior is a bit too aggressive with TS-specific syntax (namely,
+    // type annotations).
+    function getBaseContextOverride(): typeof context {
+      const reportOverride: typeof context.report = descriptor => {
+        if ('node' in descriptor && descriptor.loc == null) {
+          const { node, ...rest } = descriptor;
+          // We only want to special case the report loc when reporting on
+          // variables declarations that are not initialized. Declarations that
+          // _are_ initialized get reported by the base rule due to a setting to
+          // prohibit initializing variables entirely, in which case underlining
+          // the whole node including the type annotation and initializer is
+          // appropriate.
+          if (
+            node.type === AST_NODE_TYPES.VariableDeclarator &&
+            node.init == null
+          ) {
+            context.report({
+              ...rest,
+              loc: getReportLoc(node),
+            });
+            return;
+          }
+        }
+
+        context.report(descriptor);
+      };
+
+      // `return { ...context, report: reportOverride }` isn't safe because the
+      // `context` object has some getters that need to be preserved.
+      //
+      // `return new Proxy(context, ...)` doesn't work because `context` has
+      // non-configurable properties that throw when constructing a Proxy.
+      //
+      // So, we'll just use Proxy on a dummy object and use the `get` trap to
+      // proxy `context`'s properties.
+      return new Proxy({} as typeof context, {
+        get: (target, prop, receiver): unknown =>
+          prop === 'report'
+            ? reportOverride
+            : Reflect.get(context, prop, receiver),
+      });
+    }
```

Comment:
😬 It's a little awkward having to `Proxy` the base rule. Is there an issue that was ever filed on ESLint about better handling of report locations in this case? I couldn't find one. But even if it's declined as out-of-scope, it'd be good for reference.

Reply:
I'm not sure... I basically copied this pattern from 

https://github.com/typescript-eslint/typescript-eslint/blob/21375f65c734afc4545c0c3e7a15f72d3959c660/packages/eslint-plugin/src/rules/lines-around-comment.ts#L417-L430

and 

https://github.com/typescript-eslint/typescript-eslint/blob/3c6379b7678bcb190ae70d211cb3930c942d17a0/packages/eslint-plugin/src/rules/prefer-destructuring.ts#L196-L206

And modified it with stylistic preferences. 

Following the trail through the git history, it doesn't look like any investigation into alternatives or issues with eslint were documented in those.

(breadcrumbs)
  - https://github.com/typescript-eslint/typescript-eslint/pull/5327: first appearance of proxy pattern, not remarked upon.
  - https://github.com/typescript-eslint/typescript-eslint/pull/7117: pattern copied, remark in https://github.com/typescript-eslint/typescript-eslint/pull/7117#discussion_r1285109350 addresses caching of the proxy, but not the proxy itself.

Will plan to check for eslint issues on the topic before this PR gets merged

Reply:
> before this PR gets merged

Oh @kirkwaiblinger I was going to merge today for the Monday release and consider this a followup. But in that case I'll just let you merge whenever you want.

Reply:
Followup makes sense to me too! 👍
Happy for us to merge for the release

Reply:
Opened https://github.com/eslint/eslint/discussions/18557